File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    answer = 0
    
    if target not in words:
        return answer
    
    isvisited = [False] * len(words)
    tmp = []
    def dfs(i, begin, target, words, cnt, before):       
        if check(before, begin) > 1:
            return
        
        if begin == target:
            tmp.append(cnt)
            return

        
        for j in range(len(words)):
            if isvisited[j]: continue
            isvisited[j] = True
            dfs(j, words[j], target, words, cnt + 1, begin)
            isvisited[j] = False
    
    def check(b, a):
        cnt = 0
        for i in range(len(b)):
            if b[i] != a[i]:
                cnt += 1
        return cnt
        
    
    dfs(0, begin, target, words, 0, '')
    answer = min(tmp)
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    const visited = [];
    const q = [];
    if(!words.includes(target)){
        return 0;
    }
    q.push([begin, answer]);
    while(q){
        let [v, cnt] = q.shift();
        if(v === target){
            return cnt;
        }
        words.forEach((word) => {
            let notEqual = 0;
            
            if(visited.includes(word)) return;
            
            for(let i = 0; i < word.length; i++){
                if(word[i] !== v[i]) notEqual ++;
            }
            
            if(notEqual === 1){
                q.push([word, ++cnt]);
                visited.push(word);
            }
        })
    }
    return answer;
}